### PR TITLE
[FLINK-27714] Migrate to java-operator-sdk v3

### PR DESCRIPTION
**Pull request by morhidi**
Hi Folks, started looking at the challenges of a potential java-operator-sdk v3 migration. The main motivation for the upgrade is to enable:

- [Dynamic namespaces](https://github.com/java-operator-sdk/java-operator-sdk/issues/1154) (without operator restart)
- Embrace v3 features that simplifies the current logic (secondary resources / error handling / status patch / controller configs / etc.)

This PR contains basically the necessary changes for the v3 upgrade:
- Using operator config overrides with built in `ConfigurationServiceOverrider`
- Unify the EventSource implementations in a dedicated `EventSourceUtility`
- The `InformerManager` based secondary resource lookup was replaced by the built-in functionality in the Operator 
- The InformerManager is still used in the Webhook, due to lack of framework support in that part
- Minor changes to handle refactored functionality between v2 and v3
- Minor unit test changes

cc @wangyang0918 @Aitozi @tweise Trying to create just a working version first without too many changes.

### CODE REVIEW

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkDeploymentController.java
Comment (csviri) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkDeploymentController.java:199]: Now additional indexer is supported for this, so there is no need to traverse the list of objects, see:
https://github.com/java-operator-sdk/java-operator-sdk/blob/5afdecd3a00b2a5d1c0e6170f4c942bda3126dc8/operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/informer/ManagedInformerEventSource.java#L124-L124

Comment (csviri) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkDeploymentController.java:199]: see sample here: https://github.com/java-operator-sdk/java-operator-sdk/blob/220c79b6a5942929dc0af9a162e90c648e0f49d0/operator-framework/src/test/java/io/javaoperatorsdk/operator/sample/primaryindexer/PrimaryIndexerTestReconciler.java#L16-L16

Comment (morhidi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkDeploymentController.java]: thx!

**Comment by Aitozi** [comment]
Thanks @morhidi for this nice work, I'll take a closer look today

**Comment by morhidi** [comment]
fyi @Aitozi I run into an issue while testing with `basic-checkpoint-ha.yaml`. Once you restart the operator the Flink Session Jobs won't find their FlinkDeployment anymore. This is a bug in the java-operator-sdk. @csviri was kind enough to help me out. Quickly identified the RC https://github.com/java-operator-sdk/java-operator-sdk/issues/1238 and already fixed it https://github.com/java-operator-sdk/java-operator-sdk/pull/1237. A patch version is coming soon.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/deployment/AbstractDeploymentObserver.java
Comment (Aitozi) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/deployment/AbstractDeploymentObserver.java:202]: unused `flinkApp`, since it is a one line method call, do we still need a method ?

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java
Comment (morhidi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java:66]: Correct, that's a larger change, will open a separate ticket for that. I tried to keep the current PR as minimal as possible.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/deployment/AbstractDeploymentObserver.java
Comment (morhidi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/deployment/AbstractDeploymentObserver.java:202]: No, I removed it most of the places.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/informer/InformerManager.java
Comment (morhidi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/informer/InformerManager.java:58]: Yes, you're correct

**Review by Aitozi — APPROVED** [review]
Looks good!

**Comment by morhidi** [comment]
> Getting the following issue intermittently on submitting the `basic-session-job.yaml`
> 
> ```
> 2022-05-24 14:17:29,863 i.j.o.p.e.ReconciliationDispatcher [ERROR] [default.basic-session-job-example] Error during event processing ExecutionScope{ resource id: ResourceID{name='basic-session-job-example', namespace='default'}, version: 853952} failed.
> io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: PUT at: https://127.0.0.1:55618/apis/flink.apache.org/v1beta1/namespaces/default/flinksessionjobs/basic-session-job-example. Message: Operation cannot be fulfilled on flinksessionjobs.flink.apache.org "basic-session-job-example": the object has been modified; please apply your changes to the latest version and try again. Received status: Status(apiVersion=v1, code=409, details=StatusDetails(causes=[], group=flink.apache.org, kind=flinksessionjobs, name=basic-session-job-example, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=Operation cannot be fulfilled on flinksessionjobs.flink.apache.org "basic-session-job-example": the object has been modified; please apply your changes to the latest version and try again, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=Conflict, status=Failure, additionalProperties={}).
> 	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.requestFailure(OperationSupport.java:682)
> 	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.requestFailure(OperationSupport.java:661)
> 	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.assertResponseCode(OperationSupport.java:612)
> 	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:555)
> 	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:518)
> 	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleUpdate(OperationSupport.java:342)
> 	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleUpdate(OperationSupport.java:322)
> 	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.handleUpdate(BaseOperation.java:649)
> 	at io.fabric8.kubernetes.client.dsl.base.HasMetadataOperation.lambda$replace$1(HasMetadataOperation.java:195)
> 	at io.fabric8.kubernetes.client.dsl.base.HasMetadataOperation.replace(HasMetadataOperation.java:200)
> 	at io.fabric8.kubernetes.client.dsl.base.HasMetadataOperation.replace(HasMetadataOperation.java:141)
> 	at io.fabric8.kubernetes.client.dsl.base.HasMetadataOperation.replace(HasMetadataOperation.java:43)
> 	at io.javaoperatorsdk.operator.processing.event.ReconciliationDispatcher$CustomResourceFacade.replaceResourceWithLock(ReconciliationDispatcher.java:344)
> 	at io.javaoperatorsdk.operator.processing.event.ReconciliationDispatcher.updateCustomResourceWithFinalizer(ReconciliationDispatcher.java:302)
> 	at io.javaoperatorsdk.operator.processing.event.ReconciliationDispatcher.handleReconcile(ReconciliationDispatcher.java:101)
> 	at io.javaoperatorsdk.operator.processing.event.ReconciliationDispatcher.handleDispatch(ReconciliationDispatcher.java:80)
> 	at io.javaoperatorsdk.operator.processing.event.ReconciliationDispatcher.handleExecution(ReconciliationDispatcher.java:55)
> 	at io.javaoperatorsdk.operator.processing.event.EventProcessor$ControllerExecution.run(EventProcessor.java:356)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
> 	at java.base/java.lang.Thread.run(Thread.java:829)
> ```
> 
> @csviri suggested to remove the status initialization, but the issue still persist.
> 
> ```
> protected FlinkSessionJobStatus initStatus() {
>         return new FlinkSessionJobStatus();
>  }
> ```

There's a fix for this already https://github.com/java-operator-sdk/java-operator-sdk/pull/1244

**Comment by morhidi** [comment]
We have identified a couple of additional corner cases with @csviri 
https://github.com/java-operator-sdk/java-operator-sdk/pull/1243 <- fixed
https://github.com/java-operator-sdk/java-operator-sdk/issues/1245 <- expected to be fixed by https://github.com/java-operator-sdk/java-operator-sdk/issues/1233

**Comment by morhidi** [comment]
None of the aforementioned issues are blockers, and the retry mechanism in JOSDK solves them under the hood, but we can wait for another patch/minor release before merging to be completely on the safe side. According to @csviri a new version of JOSDK containing these fixes is expected to be released within 1-2 days.

**Comment by gyfora** [comment]
+1 for waiting 1-2 days for the JOSDK fixes before merging

**Review by gyfora — APPROVED** [review]

**Comment by morhidi** [comment]
Bumped the JOSDK version to 3.0.2

**Comment by morhidi** [comment]
I haven't hit any issues while smoke testing. Thanks @csviri for quickly addressing our findings.

**Pull request merged**
